File: sdks/go/types/conversation.go
package types

import "time"

// ConversationStatus type
type ConversationStatus string

const (
	ConversationStatusOpen     ConversationStatus = "open"
	ConversationStatusPending  ConversationStatus = "pending"
	ConversationStatusResolved ConversationStatus = "resolved"
	ConversationStatusSnoozed  ConversationStatus = "snoozed"
)

// SenderType type
type SenderType string

const (
	SenderTypeContact SenderType = "contact"
	SenderTypeAgent   SenderType = "agent"
	SenderTypeBot     SenderType = "bot"
	SenderTypeSystem  SenderType = "system"
)

// MediaContent in messages
type MediaContent struct {
	URL          string `json:"url"`
	MimeType     string `json:"mimeType"`
	Filename     string `json:"filename,omitempty"`
	Size         int64  `json:"size,omitempty"`
	Caption      string `json:"caption,omitempty"`
	ThumbnailURL string `json:"thumbnailUrl,omitempty"`
}

// LocationContent in messages
type LocationContent struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Name      string  `json:"name,omitempty"`
	Address   string  `json:"address,omitempty"`
}

// ButtonContent in messages
type ButtonContent struct {
	Type    string `json:"type"` // reply, url, call
	Text    string `json:"text"`
	Payload string `json:"payload,omitempty"`
	URL     string `json:"url,omitempty"`
	Phone   string `json:"phone,omitempty"`
}

// MessageContent structure
type MessageContent struct {
	Text     string           `json:"text,omitempty"`
	Media    *MediaContent    `json:"media,omitempty"`
	Location *LocationContent `json:"location,omitempty"`
	Buttons  []ButtonContent  `json:"buttons,omitempty"`
}

// Message model
type Message struct {
	ID             string                 `json:"id"`
	ConversationID string                 `json:"conversationId"`
	Direction      MessageDirection       `json:"direction"`
	ContentType    ContentType            `json:"contentType"`
	Content        MessageContent         `json:"content"`
	Status         MessageStatus          `json:"status"`
	ExternalID     string                 `json:"externalId,omitempty"`
	SenderID       string                 `json:"senderId,omitempty"`
	SenderType     SenderType             `json:"senderType"`
	DeliveredAt    *time.Time             `json:"deliveredAt,omitempty"`
	ReadAt         *time.Time             `json:"readAt,omitempty"`
	FailedReason   string                 `json:"failedReason,omitempty"`
	Metadata       map[string]interface{} `json:"metadata,omitempty"`
	Timestamps
}

// Conversation model
type Conversation struct {
	ID            string                 `json:"id"`
	TenantID      string                 `json:"tenantId"`
	ChannelID     string                 `json:"channelId"`
	ChannelType   ChannelType            `json:"channelType"`
	ContactID     string                 `json:"contactId"`
	Status        ConversationStatus     `json:"status"`
	AssignedTo    string                 `json:"assignedTo,omitempty"`
	AssignedAt    *time.Time             `json:"assignedAt,omitempty"`
	LastMessage   *Message               `json:"lastMessage,omitempty"`
	LastMessageAt *time.Time             `json:"lastMessageAt,omitempty"`
	UnreadCount   int                    `json:"unreadCount"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
	Tags          []string               `json:"tags,omitempty"`
	Timestamps
}

// ListConversationsParams for list request
type ListConversationsParams struct {
	PaginationParams
	Status      ConversationStatus `json:"status,omitempty"`
	ChannelID   string             `json:"channelId,omitempty"`
	ChannelType ChannelType        `json:"channelType,omitempty"`
	AssignedTo  string             `json:"assignedTo,omitempty"`
	ContactID   string             `json:"contactId,omitempty"`
	Tags        []string           `json:"tags,omitempty"`
	Search      string             `json:"search,omitempty"`
	SortBy      string             `json:"sortBy,omitempty"`
	SortOrder   string             `json:"sortOrder,omitempty"`
}

// SendMessageInput for sending messages
type SendMessageInput struct {
	Text     string                 `json:"text,omitempty"`
	Media    *MediaContent          `json:"media,omitempty"`
	Location *LocationContent       `json:"location,omitempty"`
	Buttons  []ButtonContent        `json:"buttons,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// UpdateConversationInput for updating conversations
type UpdateConversationInput struct {
	Status     ConversationStatus     `json:"status,omitempty"`
	AssignedTo *string                `json:"assignedTo,omitempty"`
	Tags       []string               `json:"tags,omitempty"`
	Metadata   map[string]interface{} `json:"metadata,omitempty"`
}

File: sdks/typescript/src/types/conversation.ts
/**
 * Conversation and Message types
 */

import type {
  ChannelType,
  ContentType,
  MessageDirection,
  MessageStatus,
  PaginationParams,
  Timestamps,
} from './common';
import type { Contact } from './contact';

export interface Conversation extends Timestamps {
  id: string;
  tenantId: string;
  channelId: string;
  channelType: ChannelType;
  contactId: string;
  contact?: Contact;
  status: ConversationStatus;
  assignedTo?: string;
  assignedAt?: string;
  lastMessage?: Message;
  lastMessageAt?: string;
  unreadCount: number;
  metadata?: Record<string, unknown>;
  tags?: string[];
}

export type ConversationStatus = 'open' | 'pending' | 'resolved' | 'snoozed';

export interface Message extends Timestamps {
  id: string;
  conversationId: string;
  direction: MessageDirection;
  contentType: ContentType;
  content: MessageContent;
  status: MessageStatus;
  externalId?: string;
  senderId?: string;
  senderType: 'contact' | 'agent' | 'bot' | 'system';
  deliveredAt?: string;
  readAt?: string;
  failedReason?: string;
  metadata?: Record<string, unknown>;
}

export interface MessageContent {
  text?: string;
  media?: MediaContent;
  location?: LocationContent;
  contact?: ContactContent;
  buttons?: ButtonContent[];
  listSections?: ListSection[];
  template?: TemplateContent;
}

export interface MediaContent {
  url: string;
  mimeType: string;
  filename?: string;
  size?: number;
  caption?: string;
  thumbnailUrl?: string;
}

export interface LocationContent {
  latitude: number;
  longitude: number;
  name?: string;
  address?: string;
}

export interface ContactContent {
  name: string;
  phones?: string[];
  emails?: string[];
}

export interface ButtonContent {
  type: 'reply' | 'url' | 'call';
  text: string;
  payload?: string;
  url?: string;
  phone?: string;
}

export interface ListSection {
  title: string;
  rows: ListRow[];
}

export interface ListRow {
  id: string;
  title: string;
  description?: string;
}

export interface TemplateContent {
  name: string;
  language: string;
  components?: TemplateComponent[];
}

export interface TemplateComponent {
  type: 'header' | 'body' | 'button';
  parameters?: TemplateParameter[];
}

export interface TemplateParameter {
  type: 'text' | 'image' | 'document' | 'video';
  text?: string;
  media?: MediaContent;
}

// Request types
export interface ListConversationsParams extends PaginationParams {
  status?: ConversationStatus;
  channelId?: string;
  channelType?: ChannelType;
  assignedTo?: string;
  contactId?: string;
  tags?: string[];
  search?: string;
  sortBy?: 'lastMessageAt' | 'createdAt' | 'updatedAt';
  sortOrder?: 'asc' | 'desc';
}

export interface SendMessageInput {
  text?: string;
  media?: {
    type: ContentType;
    url: string;
    filename?: string;
    caption?: string;
  };
  location?: LocationContent;
  buttons?: ButtonContent[];
  template?: TemplateContent;
  metadata?: Record<string, unknown>;
}

export interface UpdateConversationInput {
  status?: ConversationStatus;
  assignedTo?: string | null;
  tags?: string[];
  metadata?: Record<string, unknown>;
}

export interface ListMessagesParams extends PaginationParams {
  before?: string;
  after?: string;
}
